### PR TITLE
rc: Add operations/uploadfile to upload a file through rc using multipart/form-data

### LINKED ISSUE #3412: Upload files through the rc
#### What problem are you are trying to solve?
The rcd REST api is useful for web clients. I am working on a javascript interface for browsers.
Currently there is no way to transfer data directly from the browser to a remote.

#### How do you think rclone should be changed to solve that?
It would be very useful if the cat and rcat options were given their own api endpoints.
rcat would need to accept multipart/form-data for large uploads.
cat would be good to support the `Range`, `Content-Range`, and `Content-Length` headers.

### PR DESCRIPTION
**Pull request by negative0**
#### What is the purpose of this change?
Add functionality to upload a file to a remote using the rc. 
This will allow a file to be uploaded directly through the web GUI.
May close #3412

#### Was the change discussed in an issue or in the forum before?
No

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

**Comment by calisro** [comment]
Interesting.  I was actually considering doing this very thing in the httplib so that we could optionally upload files from the serve webdav/http templates.

**Comment by negative0** [comment]
> Interesting. I was actually considering doing this very thing in the httplib so that we could optionally upload files from the serve webdav/http templates.

Will this be of help to you?

**Comment by calisro** [comment]
@negative0 I believe so!  When I have a chance i'm going to see what needs to be done on the http/webdav side but this may help me expedite that.  Thanks!

**Comment by negative0** [comment]
Nice @calisro . How to fix the failing race tests?

**Comment by calisro** [comment]
I believe it may be inefficient to use a application/x-www-form-urlencoded to POST it which is, I believe, the option you're talking about.  It would work but for binary data I believe each non-alpha character gets encoded by the browser.  Thats 3:1 for each.

**Comment by ncw** [comment]
The problem with multipart uploads is they are quite difficult to stream (but not impossible - we do this in some of the rclone backends).

The actual file data is sent as binary data so it is efficient, but there are mime markers bounding it

Here is a really clear explanation of what actually gets sent: https://ec.haxx.se/http/http-multipart

Doing a bit more reading I think we need to call the [Request.MultipartReader](https://golang.org/pkg/net/http/#Request.MultipartReader) which makes receiving multipart uploads as a stream easy on the server.

So probably what we should do is put the entire http.Request into the input rc.Params and stream the input files to `operations.Rcat` like [this example](https://golang.org/pkg/mime/multipart/#Reader).

What do you think @negative0 ?

**Comment by negative0** [comment]
> So probably what we should do is put the entire http.Request into the input rc.Params and stream the input files to operations.Rcat like this example.

@ncw, This is an interesting approach. I will try to do that.

**Comment by negative0** [comment]
Should we add the request into the params only if it's a ```multipart/formdata``` or regardless?

**Comment by ncw** [comment]
> Should we add the request into the params only if it's a `multipart/formdata` or regardless?

A good question...

Maybe we should add the request in always with an `_request` input? That will break some things to for example the `core/echo` which is used for testing...

I suggest you try it and see how many tests break!

**Comment by negative0** [comment]
Interesting :-p

**Comment by ncw** [comment]
> Interesting :-p

What we should probably do is add a new flag here

https://github.com/rclone/rclone/blob/50e31c6636bfec69da5ad995a47839d5d4e02be9/fs/rc/registry.go#L19-L25

Called `NeedsRequest` and only add the `_request` parameter it on rc methods which have that set.

That should avoid the suprises!

What do you think?

**Comment by ncw** [comment]
The needs request code looks great :-)

How are you at re-writing history?

Can you re-arrange these commits into just two commits, so the first one is your needs request code and the second one is the upload file?

**Comment by negative0** [comment]
> 
> 
> The needs request code looks great :-)
> 
> How are you at re-writing history?
> 
> Can you re-arrange these commits into just two commits, so the first one is your needs request code and the second one is the upload file?

I'll do that :-)

**Comment by negative0** [comment]
Done :)

File: fs/rc/params.go [hunk 1]
```diff
@@ -90,6 +91,23 @@ func (p Params) Get(key string) (interface{}, error) {
...
+// ErrParamNotFound and the returned value will be nil.
+func (p Params) GetHTTPRequest() (*http.Request, error) {
+	key := "_request"
+	value, err := p.Get(key)
+	if err != nil {
+		return nil, err
+	}
+	request, ok := value.(*http.Request)
+	if !ok {
+		return nil, ErrParamInvalid{errors.Errorf("expecting file value for key %q (was %T)", key, value)}
```
[the remark this answers is not in the document]
Comment (negative0) [reply fs/rc/params.go:106]: Ahh, Yes.

**Comment by negative0** [comment]
@ncw This PR is ready to be merged!

**Comment by negative0** [comment]
I am not very good at writing tests. I will give it my best!. 

Expect updates this week.

**Comment by negative0** [comment]
Thank you :-)

**Pull request merged**